common/network-types/src/frame.rs

```diff
@@ -0,0 +1,315 @@
+        Self {
+            frame_id: initial.frame_id,
+            segments: (0..initial.seq_len)
+                .map(|i| {
+                    if i == initial.seq_idx {
+                        OnceLock::from(initial.data)
+                    } else {
+                        OnceLock::new()
+                    }
+                })
+                .collect(),
+            missing: AtomicU16::new(initial.seq_len - 1),
+        }
+    }
+
+    /// Adds a new segment to the builder.
+    fn put(&self, frame: Segment<'a>) -> u16 {
+        if !self.is_complete() {
+            if self.segments[frame.seq_idx as usize].set(frame.data).is_ok() {
+                self.missing.fetch_sub(1, Ordering::SeqCst);
+            }
+            self.missing.load(Ordering::SeqCst)
+        } else {
+            0
+        }
+    }
+
+    /// Checks if the builder contains all segments of the frame.
+    fn is_complete(&self) -> bool {
+        self.missing.load(Ordering::SeqCst) == 0
+    }
+
+    /// Reassembles the [Frame]. Panics if not [complete](FrameBuilder::is_complete).
+    fn reassemble(self) -> Frame {
+        assert!(self.is_complete(), "missing frames");
+        Frame {
+            frame_id: self.frame_id,
+            data: self
+                .segments
+                .into_iter()
+                .map(|lock| lock.into_inner().unwrap())
+                .collect::<Vec<&[u8]>>()
+                .concat()
+                .into_boxed_slice(),
+        }
+    }
+}
+
+/// Represents a frame reassembler.
+/// The [FrameReassembler] behaves as a sink [Sink] for [Segment].
+/// Upon creation, also [Stream] for reassembled [Frames](Frame) is created.
+/// The corresponding stream is closed either when [FrameReassembler::close] or
+/// [futures::SinkExt::close] is called.
+///
+/// ````rust
+/// # futures::executor::block_on(async {
+/// use hopr_network_types::frame::{Frame, FrameReassembler};
+/// use futures::{pin_mut, StreamExt};
+///
+/// let bytes = b"deadbeefcafe00112233";
+///
+/// // Build Frame and segment it
+/// let frame = Frame { frame_id: 1, data: bytes.as_ref().into() };
+/// let segments = frame.segment(2);
+/// assert_eq!(bytes.len() / 2, segments.len());
+///
+/// // Create FrameReassembler and feed the segments to it
+/// let (fragmented, reassembled) = FrameReassembler::new();
+/// pin_mut!(reassembled);
+///
+/// for segment in segments {
+///     fragmented.push_segment(segment).unwrap();
+/// }
+/// fragmented.close();
+///
+/// assert_eq!(Some(frame), reassembled.next().await);
+/// # });
+/// ````
+#[derive(Debug)]
+pub struct FrameReassembler<'a> {
+    sequences: DashMap<u32, FrameBuilder<'a>>,
+    earliest_frame: AtomicU32,
+    reassembled: async_channel::Sender<Frame>,
+}
+
+impl<'a> FrameReassembler<'a> {
+    /// Creates a new frame reassembler and a corresponding stream
+    /// for reassembled [Frames](Frame).
+    pub fn new() -> (Self, impl Stream<Item = Frame>) {
```

Comment:
This would be the best change actually, as it would templetize the reassembler and a more generic stream extension could be created for arbitrary reassembling operations. E.g. `ReassemblingStreamExt`.

Reply:
The generalization would enforce more changes though, .e.g async approach and `Send` `Sync`.

Reply:
Good idea.

Reply:
True, but the code can be organized so that there are these sync components,... and only the ext could be async.

Reply:
Since `Sink::send` or `Sink::feed` take `&mut self` it will even force the `push_segment` method to be also `&mut self` which is unwanted (because the entire reassembler is MPSC).